File: src/main/scala/com/today36524/fibonacci/Fibonacci.scala
package com.today36524.fibonacci

import scala.annotation.tailrec

/**
  * 斐波那契数 (古典兔子：1,1,2,3,5,8,13,21...)
  */
class Fibonacci {


  @tailrec
  private def fibonacciRec(remain: Int, n1: Long = 1, n2: Long = 1): Long = {
    remain match {
      case 0 => println(n1);n1
      case _ => println(n1)
        fibonacciRec(remain - 1, n2, n1 + n2)
    }
  }

  def fibonacci(remain:Int = 10): Long = {
    fibonacciRec(remain = remain)
  }

}

File: src/main/java/com/herohoy/fibonacci/Fibonacci.java
package com.herohoy.fibonacci;

import com.herohoy.tailrec.TailInvoke;
import com.herohoy.tailrec.TailRecursion;

/**
 * @author lihui  2018/3/26 下午5:46
 **/
public class Fibonacci {
    private static TailRecursion<Long> fibonacciRec (int remain,final Long n1,final Long n2) {
        if(remain == 0){
            System.out.println(n1);
            return TailInvoke.done(n1);
        }else{
            System.out.println(n1);
            return TailInvoke.call(() -> fibonacciRec(remain - 1,n2,n1+n2));
        }
    }

    public static Long fibonacci(int remain){
        remain = remain < 0 ? 10 : remain;
        return fibonacciRec(remain,1L,1L).invoke();
    }
}
